diff --git a/nixos/modules/services/cluster/k3s/default.nix b/nixos/modules/services/cluster/k3s/default.nix
index fe73c178d491..aa7d3d5e41a6 100644
--- a/nixos/modules/services/cluster/k3s/default.nix
+++ b/nixos/modules/services/cluster/k3s/default.nix
@@ -574,6 +574,14 @@ in
       '';
     };
 
+    nofile = lib.mkOption {
+      type = lib.types.ints.positive;
+      default = 1048576;
+      description = ''
+        Limit of number of file descriptors.
+      '';
+    };
+
     charts = lib.mkOption {
       type = with lib.types; attrsOf (either path package);
       default = { };
@@ -875,7 +883,7 @@ in
           Delegate = "yes";
           Restart = "always";
           RestartSec = "5s";
-          LimitNOFILE = 1048576;
+          LimitNOFILE = cfg.nofile;
           LimitNPROC = "infinity";
           LimitCORE = "infinity";
           TasksMax = "infinity";
